generator/overrides/gen.go

```diff
@@ -0,0 +1,435 @@
+package overrides
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/format"
+
+	//	"go/format"
+	"go/printer"
+	"go/token"
+	"io"
+	"regexp"
+	"strings"
+
+	"golang.org/x/tools/go/ast/astutil"
+
+	"github.com/devfile/api/generator/genutils"
+	"github.com/go-toolsmith/astcopy"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+
+	"github.com/elliotchance/orderedmap"
+)
+
+//go:generate go run sigs.k8s.io/controller-tools/cmd/helpgen generate:headerFile=../header.go.txt,year=2020 paths=.
+
+// +controllertools:marker:generateHelp:category=Overrides
+
+// FieldOverridesInclude drives whether a field should be overriden in devfile parent or plugins
+type FieldOverridesInclude struct {
+	// Omit indicates that this field cannot be overriden in a devfile plugin.
+	Omit bool `marker:",optional"`
+	// OmmitInPlugin indicates that this field cannot be overriden in a devfile plugin.
+	OmitInPlugin bool `marker:",optional"`
+	// Description indicates the description that should be added as Go documentation on the generated structs.
+	Description string `marker:",optional"`
+}
+
+var (
+	overridesFieldMarker = markers.Must(markers.MakeDefinition("devfile:overrides:include", markers.DescribesField, FieldOverridesInclude{}))
+	overridesTypeMarker  = markers.Must(markers.MakeDefinition("devfile:overrides:generate", markers.DescribesType, struct{}{}))
+)
+
+// +controllertools:marker:generateHelp
+
+// Generator generates additional GO code for the overriding of elements in devfile parent or plugins.
+type Generator struct {
+
+	// IsForPluginOverrides indicates that the generated code should be done for plugin overrides.
+	// When false, the parent overrides are generated
+	IsForPluginOverrides bool `marker:"isForPluginOverrides,optional"`
+
+	packageTypes      map[string]*markers.TypeInfo
+	suffix            string
+	rootTypeToProcess typeToProcess
+}
+
+// RegisterMarkers registers the markers of the Generator
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	if err := markers.RegisterAll(into, overridesFieldMarker, overridesTypeMarker); err != nil {
+		return err
+	}
+	into.AddHelp(overridesFieldMarker, FieldOverridesInclude{}.Help())
+	into.AddHelp(overridesTypeMarker, markers.SimpleHelp("Overrides", "indicates that a type should be selected to create Overrides for it"))
+	return genutils.RegisterUnionMarkers(into)
+}
+
+// Generate generates the artifacts
+func (g Generator) Generate(ctx *genall.GenerationContext) error {
+	for _, root := range ctx.Roots {
+		ctx.Checker.Check(root, func(node ast.Node) bool {
+			// ignore interfaces
+			_, isIface := node.(*ast.InterfaceType)
+			return !isIface
+		})
+
+		root.NeedTypesInfo()
+
+		var rootStructToOverride *markers.TypeInfo
+		g.packageTypes = map[string]*markers.TypeInfo{}
```

Comment:
Why ?

Reply:
Because struct fields represent internal state for the struct -- by updating it each iteration through `ctx.Roots`, we're just setting it to the last `ctx.Root`. It looks like it's being used as an instance variable. In other words, 
- If `g.packageTypes` is used elsewhere, we're setting it incorrectly since we use it to hold transient state
- If `g.packageTypes` is only used for this function, it could be just be `packageTypes := map[string]*markers.TypeInfo{}`

Reply:
Yeah, I see what you mean.

Reply:
Fixed in commit https://github.com/devfile/api/pull/133/commits/d948ac3778ab00b7872dc858e8ca750775441d56 by passing `packageTypes` as argument to functions.